### ISSUE TITLE
*_pb:v_msg_my_module/3 failed with gpb_type_error

### ISSUE THREAD

**Issue by AnufrievDE** [tomas-abrahamsson/gpb#174]
Hello, I've faced an issue that looks like a bug(or hopefully as an unexpected behavior) when verifing my message.

I've the following gpb_options enabled:
```
{gpb_opts, [{i, "protos"}, 
            {module_name_suffix, "_pb"}, 
            {o, "src"},
            {strings_as_binaries, true},
            {maps, true},
            {maps_key_type, binary},
            type_specs,
            json,
            use_packages,
            {json_format, jiffy}]}.
```

I've the following messages defined in my .proto file:
```
message my_module {
    uint32     module_id = 1;
    string     module_code = 2;
    string     module_name = 3;
    string     module_desc = 4;
    
    uint32     version_id = 5;
    string     version_code = 6;
    string     version_name = 7;
    
    //string     lang_code = 8;
    string     lang_name = 9;
}

message my_module_list {
    repeated my_module    my_module_list = 1;
}
```
Not sure that the way(with defining a separate message) I chose for representing the list of some 'other' already defined messages is the best way. But I can't imagine any reasons why it should not works well.

So relying on "{maps_key_type, binary},{maps, true}" gpb options I assume it is ok to have messages field names(maps keys) being a binaries, when preparing message for serialization. But according to *_pb:v_msg_my_module/3 impementation it seems that something wrong with my understanding or with gpb implementation or with a way I chosed to represent the list of messages as a separate message, please find an example bellow:

```
5> M2 = #{<<"lang_name">> => <<"Undefined">>,<<"module_code">> => <<"acc">>,<<"module_desc">> => <<>>, <<"module_id">> => 1, <<"module_name">> => <<208,145,209,131>>, <<"version_code">> => <<>>, <<"version_id">> => 0, <<"version_name">> => <<>>}.
#{<<"lang_name">> => <<"Undefined">>,
  <<"module_code">> => <<"acc">>,<<"module_desc">> => <<>>,
  <<"module_id">> => 1,
  <<"module_name">> => <<208,145,209,131>>,
  <<"version_code">> => <<>>,<<"version_id">> => 0,
  <<"version_name">> => <<>>}
6> configuration_pb:verify_msg(M2,my_module).
** exception error: {gpb_type_error,{{extraneous_key,<<"lang_name">>},
                                     [{value,#{<<"lang_name">> => <<"Undefined">>,
                                               <<"module_code">> => <<"acc">>,<<"module_desc">> => <<>>,
                                               <<"module_id">> => 1,
                                               <<"module_name">> => <<208,145,209,131>>,
                                               <<"version_code">> => <<>>,<<"version_id">> => 0,
                                               <<"version_name">> => <<>>}},
                                      {path,my_module}]}}
     in function  configuration_pb:mk_type_error/3 (/mnt/d/conf_server/src/configuration_pb.erl, line 8657)
     in call from lists:foreach/2 (lists.erl, line 1338)
     in call from configuration_pb:v_msg_my_module/3 (/mnt/d/conf_server/src/configuration_pb.erl, line 7798)
7> configuration_pb:verify_msg(#{my_module_list => [M2]},my_module_list). 
ok
8> configuration_pb:verify_msg(#{<<"my_module_list">> => [M2]},my_module_list).
** exception error: {gpb_type_error,{{extraneous_key,<<"lang_name">>},
                                     [{value,#{<<"lang_name">> => <<"Undefined">>,
                                               <<"module_code">> => <<"acc">>,<<"module_desc">> => <<>>,
                                               <<"module_id">> => 1,
                                               <<"module_name">> => <<208,145,209,131>>,
                                               <<"version_code">> => <<>>,<<"version_id">> => 0,
                                               <<"version_name">> => <<>>}},
                                      {path,'my_module_list.my_module_list'}]}}
     in function  configuration_pb:mk_type_error/3 (/mnt/d/conf_server/src/configuration_pb.erl, line 8657)
     in call from lists:foreach/2 (lists.erl, line 1338)
     in call from configuration_pb:v_msg_my_module/3 (/mnt/d/conf_server/src/configuration_pb.erl, line 7798)
     in call from configuration_pb:'-v_msg_my_module_list/3-lc$^0/1-0-'/3 (/mnt/d/conf_server/src/configuration_pb.erl, line 7825)
     in call from configuration_pb:v_msg_my_module_list/3 (/mnt/d/conf_server/src/configuration_pb.erl, line 7827)```

Could you help me to workaround/fix that?
Is it a bug, or I don't understand the meaning of options I use?
Maybe there are reasons why it is not recommended to use {maps_key_type, binary} option?

Thanks in advance


P.S: Issue is reproduced on
rebar3_gpb_plugin 2.12.2 (gpb 4.10.2);
rebar3_gpb_plugin 2.12.4 (gpb 4.10.4);

**Comment by tomas-abrahamsson** [comment]
Thanks for a well-written report! This looks like a bug in gpb. I'll look into it.

**Comment by AnufrievDE** [comment]
It seems I got why ok is returned in this example:
```
7> configuration_pb:verify_msg(#{my_module_list => [M2]},my_module_list). 
ok
```
We just fall into _ case clause in this case:
```
case M of
      #{<<"my_module_list">> := F1} ->      
      _ -> ok
end
```

```
v_msg_my_module_list(#{} = M, Path, TrUserData) ->
    case M of
      #{<<"my_module_list">> := F1} ->
	  if is_list(F1) ->
		 _ = [v_msg_my_module(Elem,
					 [my_module_list | Path], TrUserData)
		      || Elem <- F1],
		 ok;
	     true ->
		 mk_type_error({invalid_list_of, {msg, my_module}},
			       F1, [my_module_list | Path])
	  end;
      _ -> ok
    end,
    lists:foreach(fun (my_module_list) -> ok;
		      (OtherKey) ->
			  mk_type_error({extraneous_key, OtherKey}, M, Path)
		  end,
		  maps:keys(M)),
    ok;
...
```

Probably atoms in **lists:foreach** genereted case should be replaced with binaries when {maps_key_type,binary} option is specified.

Probably the problem with FNames = [{Type,Name,Fields}|T] where **Name** is an atom(from MsgDef) in _gpb_gen_verifiers:format_msg_verifier/3_, hopefully it should be a binary if corresponding option is enabled.
```
format_msg_verifier(MsgName, MsgDef, AnRes, Opts) ->
    FNames = gpb_lib:get_field_names(MsgDef),
```

**Comment by tomas-abrahamsson** [comment]
> Probably atoms in lists:foreach genereted case should be replaced with binaries when {maps_key_type,binary} option is specified.

Yes indeed. I just pushed a preliminary fix to the branch `fix-verifiers-for-binary-map-keys`.

**Comment by tomas-abrahamsson** [comment]
The fix has been merged into 4.10.5, which is out, so I'll close this issue. Feel free to re-open or open another, if there's anything more. (I'll remove the temporary branch in a while.)
